### PR TITLE
Separate rendering part from specification part for layers

### PR DESCRIPTION
**Pull request by ivovandongen**
In an attempt to separate rendering on a different thread..

Open ends:
- [ ] Generate parts of the code. Feels repetitious
- ~~Use of `std::enable_shared_from_this` eg [here](https://github.com/mapbox/mapbox-gl-native/blob/render-layers/src/mbgl/style/layers/background_layer_impl.hpp#L10) so we can do [this](https://github.com/mapbox/mapbox-gl-native/blob/render-layers/src/mbgl/style/layers/background_layer_impl.cpp#L8)~~
- [X] I've extracted `Layer::Type` in a [separate header](https://github.com/mapbox/mapbox-gl-native/blob/render-layers/include/mbgl/style/layer_type.hpp) to re-use it for `RenderLayer` subclasses. This makes it public though. ~~Maybe we should just duplicate it, especially if we want this to be part of the public API.~~
- [ ] Move rendering related source files to the `/renderer` package:
    - paint_property_binder
    - paint_property_statistics
    - *property_evaluator
    - ...
- [X] Thorough cleanup

cc @jfirebaugh

### CODE REVIEW

File: src/mbgl/style/layer_impl.hpp [hunk 1]
```diff
@@ -42,6 +36,7 @@ class Layer::Impl {
 
     // Create a new layer with the specified `id` and `sourceID`. All other properties
     // are copied from this layer.
+    // TODO: evaluate if this still meets expectations
```
Comment (jfirebaugh) [inline src/mbgl/style/layer_impl.hpp]: `copy` is only used in one place: `StyleSourcedAnnotationImpl`, which is code supporting a pre-runtime styling feature that we should remove. We can save that for a separate change though.

File: src/mbgl/style/style.cpp [hunk 2]
```diff
@@ -243,9 +253,57 @@ std::unique_ptr<Layer> Style::removeLayer(const std::string& id) {
...
+    auto it = std::find_if(renderLayers.begin(), renderLayers.end(), [&](const auto& layer) {
+        return layer->baseImpl->id == id;
+    });
+
+    if (it == renderLayers.end()) {
+        return;
+    }
+
+    // TODO
+//    auto layer = std::move(*it);
```
Comment (jfirebaugh) [inline src/mbgl/style/style.cpp]: This can be deleted. When a custom layer is removed, the custom layer deinitialize function only needs to be called once.

File: src/mbgl/style/layers/symbol_layer_impl.hpp [hunk 3]
```diff
@@ -9,78 +9,27 @@
...
     std::unique_ptr<Layer> cloneRef(const std::string& id) const override;
     void stringifyLayout(rapidjson::Writer<rapidjson::StringBuffer>&) const override;
 
-    void cascade(const CascadeParameters&) override;
-    bool evaluate(const PropertyEvaluationParameters&) override;
+    std::unique_ptr<RenderLayer> createRenderLayer() const override;
 
-    std::unique_ptr<Bucket> createBucket(const BucketParameters&, const std::vector<const Layer*>&) const override;
-    std::unique_ptr<SymbolLayout> createLayout(const BucketParameters&, const std::vector<const Layer*>&,
+    // TODO move to RenderSymbolLayer?
```
Comment (jfirebaugh) [inline src/mbgl/style/layers/symbol_layer_impl.hpp]: Yes, it should. This is the symbol layer equivalent of `createBucket`.

File: include/mbgl/style/layer.hpp [hunk 4]
```diff
@@ -117,14 +111,12 @@ class Layer : public mbgl::util::noncopyable {
     void setMaxZoom(float) const;
 
     // Private implementation
-    const std::unique_ptr<Impl> baseImpl;
+    const std::shared_ptr<Impl> baseImpl;
```
Comment (jfirebaugh) [inline include/mbgl/style/layer.hpp]: > `RenderLayers` should never outlive their associated `Layer`

Just realized this isn't the case. `RenderLayer`s copied into the worker may outlive the `Layer`. Need to think about this.

Comment (jfirebaugh) [inline include/mbgl/style/layer.hpp]: Per chat: good solution is probably to continue to pass copies of `Layer`s to the worker, rather than `RenderLayer`s. The worker can then create its own `RenderLayer`s, allowing us to preserve the invariant "`RenderLayers` never outlive their associated `Layer`" and avoid (unintentionally) sharing a `Layer::Impl` between the main thread and worker.

Comment (ivovandongen) [inline include/mbgl/style/layer.hpp]: @jfirebaugh If we copy layers in `GeometryTile` to pass to `GeometryTileWorker::setLayers` and then create `RenderLayer`s from that in the `GeometryTileWorker`; we would be copying the associated `Impl` twice right? In that case, maybe we should revert the changes to `GeometryTileWorker` and group_by_layout.hpp to continue to work on `Layer`s?

Comment (ivovandongen) [inline include/mbgl/style/layer.hpp]: @jfirebaugh I Fixed up ownership as discussed above in https://github.com/mapbox/mapbox-gl-native/pull/8797/commits/7f0ad3241db420e8a91242f1af7586592aa0ce47. I Left the pointers to the concrete Implementations in the subclasses as is (`const style::FooLayer::Impl* const impl;` as opposed to a `const style::FooLayer::Impl&`) to avoid modifying all call sites.

Comment (ivovandongen) [inline include/mbgl/style/layer.hpp]: @jfirebaugh This does not work as hoped. Since the `Layer::Impl` subclasses no longer contain a reference to the cascaded/evaluated paint properties, the render layers created in the `GeometryTileWorker` start with a blank copy, disjointed from the copy in the render layers in the style.

Comment (jfirebaugh) [inline include/mbgl/style/layer.hpp]: Do we actually need the cascaded/evaluated paint properties? Evaluated paint properties can be generated from the cascaded paint properties plus tile zoom level. With regards to cascaded properties, cascading would mainly be needed in supporting the combination of data-driven paint properties with style classes, which I don't think we support.

Comment (ivovandongen) [inline include/mbgl/style/layer.hpp]: @jfirebaugh I figured this must be the reason that all the property-function render tests are failing. It makes me think it's because there is some state now not copied over / held on to when creating the buckets (since that is the main change).

If I understand correctly; in [the painter](https://github.com/mapbox/mapbox-gl-native/blob/master/src/mbgl/renderer/painter_circle.cpp#L15) the paint property binders (created in the bucket from the copied paint properties) and the evaluated paint properties form the `Layer::Impl` come together and then in the program the current properties are bound to attributes and interpolated from the values the binders were originally created with right?

Comment (jfirebaugh) [inline include/mbgl/style/layer.hpp]: https://github.com/mapbox/mapbox-gl-native/commit/c397125178777b1eab5a2580c04e79552b79148a fixes the render tests.

It's true that the current paint properties are inspected at render time as well as the binder, but it's either one or the other:

* If the current value is a constant or source function, it's used and the value the binder was originally created from is ignored.
* Otherwise, the value the binder was originally created from is used, and the current value is ignored.

**Comment by ivovandongen** [comment]
@jfirebaugh I update the pr with your review comments. I refrained from moving more files to the renderer package for now as the ones I had my eye on are all referred to from `paint_properties.hpp`. Moving it to a different package would really help readability.
